### ISSUE TITLE
Different behaviour between `libc::poll` and calling `poll` in C directly on linux.

### ISSUE THREAD

**Issue by s-d-m** [rust-lang/libc#3907]
Hi,

I'm using ubuntu 20.04 via wsl and I noticed that a calling `libc::poll` and `libc` from C differ in behaviour. I bet I must have a bug in my program but can't see where.

Following code checks if stdin is closed via C code:
```
#include <poll.h>
#include <stdio.h>

int main() {
  struct pollfd poll_param = { .fd = 0, .events = 0, .revents = POLLHUP };  // checks if stdin is closed
  int res = poll(&poll_param, 1, 0);
  fprintf(stdout, "poll returned %i\n", res);
  return res;
}
```
what I thought would be the equivalent in rust is:
```
use std::ptr::addr_of_mut;
use libc::POLLHUP;

fn main() {
    let mut pollfds_input = libc::pollfd {
        fd: 0, // stdin
        events: 0,
        revents: POLLHUP,
    };
    let pollfds_input_ptr = addr_of_mut!(pollfds_input);

    let poll_res = unsafe {
        libc::poll( pollfds_input_ptr, 1, 0)
    };
    let is_stdin_closed = poll_res > 0;
    println!("is stdin closed? {is_stdin_closed}. poll_res is {poll_res}");
}
```

After compiling the c code with `gcc -o check_stdin main.c`  I can see the following:
```
bash>  ./check_stdin
poll returned 0
bash> ./check_stdin 0<&-
poll returned 1
```

The `0<&-` is bash syntax to run the command with the stdin closed. 
When running the rust program, `libc::poll` returns 0, even if I execute the command with this `0<&-`.

Is this a libc issue? Or is there something special about stdin in rust stdlib that makes this check moot from libc side?

EDIT: I can rule out the issue of not forwarding the arguments to the poll syscall. I checked with gdb by putting a breakpoint on poll and peeking at the memory at %rdi, and the data there matches the parameters at the `libc::poll` call in main

**Comment by s-d-m** [comment]
Found the issue.

The rust code calls `openat(AT_FDCWD, "/dev/null", O_RDWR)` before main starts. After this `openat` syscall, the `poll` call to check if stdin is closed or not always returns 0.

The issue is something related to a successful call to `openat` and not `/dev/null`. The rust code also calls `openat(AT_FDCWD, "/proc/self/maps", O_RDONLY|O_CLOEXEC)` before main and that shows the same behaviour here. Calling `openat` with a non existing filename allows the check for closed stdin to proceed as expected.

In other words, this isn't an issue in `libc`. 

If someone has an idea or a link on how to check if stdin is closed or not, please let me know.

**Comment by s-d-m** [comment]
Found the issue.

The rust code calls openat(AT_FDCWD, "/dev/null", O_RDWR) before main starts. After this openat syscall, the poll call to check if stdin is closed or not always returns 0.

The issue is something related to a successful call to openat and not /dev/null. The rust code also calls openat(AT_FDCWD, "/proc/self/maps", O_RDONLY|O_CLOEXEC) before main and that shows the same behaviour here. Calling openat with a non existing filename allows the check for closed stdin to proceed as expected.

In other words, this isn't an issue in libc.

If someone has an idea or a link on how to check if stdin is closed or not, please let me know.
